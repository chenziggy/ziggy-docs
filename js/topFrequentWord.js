export function splitWords(str) {
  return str.match(/[a-zA-Z0-9'-]+/g).filter(item => item).map(str => str.toLowerCase())
}

export function genHashMap(arr) {
  const map = {}
  arr.reduce((acc, word, index, arr) => {
    const prevItem = index - 1 >= 0 ? arr[index - 1] : false
    const nextItem = index + 1 < arr.length ? arr[index + 1] : false

    if (!acc[word]) {
      acc[word] = {
        value: 1,
        prev: [],
        next: [],
      }
    }
    else {
      acc[word].value++
    }
    prevItem && acc[word].prev.push(prevItem)
    nextItem && acc[word].next.push(nextItem)

    return acc
  }, map)
  return map
}

export function sortValue(hashMap) {
  return Object.keys(hashMap).sort((aKey, bKey) => {
    const a = hashMap[aKey]
    const b = hashMap[bKey]
    const cond1 = b.value - a.value
    return cond1 !== 0 ? cond1 : aKey.localeCompare(bKey)
  })
}

export function findMostFrequentString(arr) {
  if (!arr.length)
    return
  const map = {}
  arr.reduce((acc, word) => {
    if (!acc[word])
      acc[word] = { value: 1 }
    else acc[word].value++
    return acc
  }, map)
  const word = sortValue(map)[0]
  return {
    word,
    value: map[word].value,
  }
}

export function topFrequentWord(str) {
  const arr = splitWords(str)
  const hashMap = genHashMap(arr)
  const words = sortValue(hashMap).slice(0, 3)
  return words.map((word) => {
    const item = hashMap[word]
    return {
      word,
      value: item.value,
      topPrevWord: findMostFrequentString(item.prev),
      topNextWord: findMostFrequentString(item.next),
    }
  })
}
