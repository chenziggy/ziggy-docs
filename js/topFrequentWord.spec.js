import { describe, expect, it } from 'vitest'
import { findMostFrequentString, genHashMap, sortValue, splitWords, topFrequentWord } from './topFrequentWord'

describe('查找字符串中出现频率最高的3个单词，同样规则返回其相邻单词', () => {
  it('单词拆分', () => {
    const str = 'Huawei devices. company\'s (ICT)'
    expect(splitWords(str)).toEqual(['huawei', 'devices', 'company\'s', 'ict'])
  })

  it('生成hashMap', () => {
    const arr = ['huawei', 'devices', 'huawei']
    expect(genHashMap(arr)).toEqual({
      huawei: {
        value: 2,
        prev: ['devices'],
        next: ['devices'],
      },
      devices: {
        value: 1,
        prev: ['huawei'],
        next: ['huawei'],
      },
    })
  })

  it('hashMap value 大小排序 相同value a-z排序', () => {
    const hashMap = {
      devices: {
        value: 1,
        prev: ['huawei'],
        next: ['huawei'],
      },
      huawei: {
        value: 2,
        prev: ['devices'],
        next: ['devices'],
      },
      auawei: {
        value: 2,
        prev: ['devices'],
        next: ['devices'],
      },
    }
    expect(sortValue(hashMap)).toEqual(['auawei', 'huawei', 'devices'])
  })

  it('返回单词数组出现频次最高的单词及出现次数', () => {
    const words = ['auawei', 'huawei', 'devices', 'huawei']
    expect(findMostFrequentString(words)).toEqual({
      word: 'huawei',
      value: 2,
    })
  })

  it('最终组合', () => {
    const str = 'Huawei is a multinational technology company headquartered in Shenzhen, Guangdong, China. Founded in 1987, Huawei is a leading global provider of information and communications technology (ICT) infrastructure and smart devices. The company\'s portfolio includes a wide range of products and solutions, such as smartphones, tablets, laptops, telecommunications equipment, networking solutions, and cloud computing services. Huawei is known for its innovation and research in areas such as 5G technology, artificial intelligence, and consumer electronics.'

    expect(topFrequentWord(str)).toMatchSnapshot()
  })
})
