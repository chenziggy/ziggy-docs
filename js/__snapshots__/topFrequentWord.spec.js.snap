// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`查找字符串中出现频率最高的3个单词，同样规则返回其相邻单词 > 最终组合 1`] = `
[
  {
    "topNextWord": {
      "value": 1,
      "word": "cloud",
    },
    "topPrevWord": {
      "value": 1,
      "word": "information",
    },
    "value": 6,
    "word": "and",
  },
  {
    "topNextWord": {
      "value": 1,
      "word": "leading",
    },
    "topPrevWord": {
      "value": 2,
      "word": "is",
    },
    "value": 3,
    "word": "a",
  },
  {
    "topNextWord": {
      "value": 3,
      "word": "is",
    },
    "topPrevWord": {
      "value": 1,
      "word": "1987",
    },
    "value": 3,
    "word": "huawei",
  },
]
`;
